### PR TITLE
feat: make smart_delegate a global account

### PR DESCRIPTION
**Pull request by dcaf-mocha**
feat: make smart_delegate a global account
- removes close smart_delegate ix
- changes seeds for smart_delegate
- updates tests

BREAKING CHANGE: changes seeds for smart_delegate, removes
close_smart_delegate ix

### CODE REVIEW

File: program-tests/pre-authorized-debit-v1/init-smart-delegate.test.2.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    );
+
+    // verify smart delegate account
+    const smartDelegateAccount = await program.account.smartDelegate.fetch(
+      canonicalSmartDelegatePublicKey,
+    );
+    expect(smartDelegateAccount.bump).to.equal(canonicalSmartDelegateBump);
+  });
+
+  it("should fail if trying to init a smart delegate twice", async () => {
```
Comment (dcaf-mocha) [inline program-tests/pre-authorized-debit-v1/init-smart-delegate.test.2.ts:105]: A bit of an anti-pattern to have tests rely on other tests.
This test relies on the previous test to pass. 

But in the grand scheme of things it doesn't matter as all tests in this describe block run sequentially and its a simple test file.

File: program-tests/pre-authorized-debit-v1/init-smart-delegate.test.ts [hunk 2]
```diff
@@ -1,330 +0,0 @@
-import "./setup";
```
Comment (dcaf-mocha) [inline program-tests/pre-authorized-debit-v1/init-smart-delegate.test.ts:1]: File renamed + modified.

File: .github/workflows/program-tests.yml [hunk 3]
```diff
@@ -16,3 +16,15 @@ jobs:
         with:
           args: "" # add anchor test args, e.g. "--skip-lint"
           workspace_dir: "." # path to the anchor workspace, e.g. "./my_workspace"
+
+  # TODO: this is wasting some credits as it is re-compiling and re-installing packages
```
Comment (dcaf-mocha) [inline .github/workflows/program-tests.yml:20]: Created a ticket in our backlog for this

File: program-tests/pre-authorized-debit-v1/init-smart-delegate.test.2.ts (hunk 1 above)
Comment (dcaf-matcha) [reply program-tests/pre-authorized-debit-v1/init-smart-delegate.test.2.ts:105]: Yeah in this case, I think it's fine tbh

File: programs/pre-authorized-debit-v1/src/errors.rs [hunk 4]
```diff
@@ -30,8 +30,8 @@ pub enum CustomProgramError {
     #[msg("Smart delegate can only be closed by token account owner")]
     SmartDelegateCloseUnauthorized, // 6008 or 0x1778
 
-    #[msg("Smart delegate and token account mismatch")]
-    SmartDelegateTokenAccountMismatch, // 6009 or 0x1779
+    #[msg("This Error is unused")]
+    UnusedError, // 6009 or 0x1779
```
Comment (dcaf-matcha) [inline programs/pre-authorized-debit-v1/src/errors.rs:34]: Why don't you replace this with `PausePreAuthorizationUnauthorized`. You'll only have to like update one or two tests to change the error code and we don't need to leave this in here like this.

File: sdk/pre-authorized-debit-v1/package.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
     "main": "dist/index.js",
     "types": "dist/index.d.ts",
     "scripts": {
-        "build": "yarn build:anchor-client && yarn run tsc",
+        "build": "yarn run tsc",
```
Comment (dcaf-matcha) [inline sdk/pre-authorized-debit-v1/package.json:8]: Why did you remove this?

Comment (dcaf-mocha) [reply sdk/pre-authorized-debit-v1/package.json:8]: Build should ideally just produce the compiled js files. The `build:anchor-client` has a dependency on parent folders (pathing to the idl file), which is a little weird to have in a build command.

File: programs/pre-authorized-debit-v1/src/errors.rs (hunk 4 above)
Comment (dcaf-mocha) [reply programs/pre-authorized-debit-v1/src/errors.rs:34]: Good idea.

Comment (dcaf-mocha) [reply programs/pre-authorized-debit-v1/src/errors.rs]: Done.

File: sdk/pre-authorized-debit-v1/package.json (hunk 5 above)
Comment (dcaf-matcha) [reply sdk/pre-authorized-debit-v1/package.json:8]: Okay makes sense

**Review by dcaf-matcha — APPROVED** [review]

**Review by Chasewhip8 — CHANGES_REQUESTED** [review]
InitPreAuthorization needs to set the delegate and delegate amount, also it might be nice to have a bool passed into the debit ix to automatically reset to max delegated amount.

**Comment by dcaf-mocha** [comment]
> InitPreAuthorization needs to set the delegate and delegate amount if not already set, also it might be nice to have a bool passed into the debit ix to automatically reset to max delegated amount (since the token account is already mutable the only extra cost is the compute).

@dcaf-matcha I created a ticket for this. If I don't get to this by tonight, take this when you start tmrw morning.

**Comment by dcaf-matcha** [comment]
> > InitPreAuthorization needs to set the delegate and delegate amount if not already set, also it might be nice to have a bool passed into the debit ix to automatically reset to max delegated amount (since the token account is already mutable the only extra cost is the compute).
> 
> @dcaf-matcha I created a ticket for this. If I don't get to this by tonight, take this when you start tmrw morning.

sounds good, will do :+1: . And thanks for the suggestions @Chasewhip8 , really appreciate it! :pray:

**Comment by dcaf-mocha** [comment]
> InitPreAuthorization needs to set the delegate and delegate amount if not already set, also it might be nice to have a bool passed into the debit ix to automatically reset to max delegated amount (since the token account is already mutable the only extra cost is the compute).


https://github.com/seabed-labs/pre-authorized-debit/pull/38

> have a bool passed into the debit ix to automatically reset to max delegated amount
This is not possible since the token_account.owner does not sign the `debit` instructions, and delegates do not have the authority to change their own delegated amount.
